urls::string_token::arg struct has virtual functions but not virtual destructor

On MSVC 17, Using this library git tag `boost-1.81.0.beta1`, boost-1.80.0 (installed via conan) and this library (installed with CMake-FetchContent) I ran into a compiler error when I had elevated warning [C4265 Virtual class missing virtual destructor](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4265?view=msvc-170) to error. 

main.cpp
```
#include <boost/url/url.hpp>

int main()
{
  boost::urls::url url("http://boost.org");
  return 0;
}
```

CMakeLists.txt
```
CMAKE_MINIMUM_REQUIRED(VERSION 3.15)

project(use_boost_url)

find_package(Boost REQUIRED)
include(FetchContent)
FetchContent_Declare(boost_url_copy
  GIT_REPOSITORY https://github.com/boostorg/url.git
  GIT_TAG boost-1.81.0.beta1
  GIT_SHALLOW 1
  )
FetchContent_MakeAvailable(boost_url_copy)

add_library(demo main.cpp)
target_link_libraries(demo Boost::url)

# promote virtual class without virtual destructor to error
# https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4265?view=msvc-170
target_compile_options(demo PUBLIC /we4265)
```
conanfile.txt
```
[requires]
boost/1.80.0

[generators]
CMakeDeps
CMakeToolchain
````

conan install and cmake configure including some output to show compiler version
```
>mkdir build && cd build
>conan install .. -s build_type=Release -s compiler="Visual Studio" -s compiler.version=17 -s arch=x86_64 --build=missing
>cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake 
-- Building for: Visual Studio 17 2022
-- Using Conan toolchain: C:/Users/User/repos/use_boost_url/build/conan_toolchain.cmake
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19043.
-- The C compiler identification is MSVC 19.33.31630.0
-- The CXX compiler identification is MSVC 19.33.31630.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.33.31629/bin/Hostx64/x64/cl.exe - skipped
```

Build output
```
\build>cmake --build . --config Release
MSBuild version 17.3.1+2badb37d1 for .NET Framework
  Checking File Globs
  Checking Build System
  Building Custom Rule C:/Users/User/repos/use_boost_url/build/_deps/boost_url_copy-src/CMakeLists.txt
  src.cpp
  boost_url.vcxproj -> C:\Users\User\repos\use_boost_url\build\_deps\boost_url_copy-build\Release\boost_url.lib
  Building Custom Rule C:/Users/User/repos/use_boost_url/CMakeLists.txt
  main.cpp
C:\Users\User\repos\use_boost_url\build\_deps\boost_url_copy-src\include\boost/url/grammar/string_token.hpp(137,1): error C4265: 'boost::ur
ls::string_token::return_string': class has virtual functions, but its non-trivial destructor is not virtual; instances of this class may n
ot be destructed correctly [C:\Users\User\repos\use_boost_url\build\demo.vcxproj]
```

--- a/extra/test_main.cpp
+++ b/extra/test_main.cpp
@@ -133,6 +133,8 @@ instance() noexcept
         std::vector<any_suite const*> v_;
 
     public:
+        virtual ~suites_impl() = default;
+
         void
         insert(any_suite const& t) override
         {
@@ -375,7 +377,7 @@ public:
         v_.reserve(256);
     }
 
-    ~simple_runner()
+    virtual ~simple_runner()
     {
         log_ <<
             elapsed{clock_type::now() -

--- a/extra/test_suite.hpp
+++ b/extra/test_suite.hpp
@@ -43,6 +43,8 @@ struct any_suite
 
 struct suites
 {
+    virtual ~suites() = default;
+
     using iterator = any_suite const* const*;
     virtual void insert(any_suite const&) = 0;
     virtual iterator begin() const noexcept = 0;
@@ -95,7 +97,7 @@ public:
     static any_runner& instance() noexcept;
 
     any_runner() noexcept;
-    ~any_runner();
+    virtual ~any_runner();
 
     virtual void run(any_suite const& test) = 0;
     virtual void note(char const* msg) = 0;

--- a/include/boost/url/detail/any_segments_iter.hpp
+++ b/include/boost/url/detail/any_segments_iter.hpp
@@ -31,6 +31,8 @@ protected:
     {
     }
 
+    virtual ~any_segments_iter() = default;
+
 public:
     // this is adjusted
     // when self-intersecting
@@ -76,6 +78,8 @@ struct BOOST_SYMBOL_VISIBLE
     path_iter
     : any_segments_iter
 {
+    virtual ~path_iter() = default;
+
     explicit
     path_iter(
         string_view s) noexcept;
@@ -97,6 +101,8 @@ struct BOOST_SYMBOL_VISIBLE
     path_encoded_iter
     : public path_iter
 {
+    virtual ~path_encoded_iter() = default;
+
     explicit
     path_encoded_iter(
         pct_string_view s) noexcept;
@@ -118,6 +124,8 @@ struct BOOST_SYMBOL_VISIBLE
     segment_iter
     : any_segments_iter
 {
+    virtual ~segment_iter() = default;
+
     explicit
     segment_iter(
         string_view s) noexcept;
@@ -228,6 +236,8 @@ struct BOOST_SYMBOL_VISIBLE
     segment_encoded_iter
     : any_segments_iter
 {
+    virtual ~segment_encoded_iter() = default;
+
     explicit
     segment_encoded_iter(
         pct_string_view const& s) noexcept;

--- a/include/boost/url/grammar/impl/error.ipp
+++ b/include/boost/url/grammar/impl/error.ipp
@@ -22,6 +22,8 @@ make_error_code(
 {
     struct codes : error_category
     {
+        virtual ~codes() = default;
+
         codes() noexcept
             : error_category(
                 0x0536e50a30f9e9f2)
@@ -77,6 +79,8 @@ make_error_condition(
 {
     struct codes : error_category
     {
+        virtual ~codes() = default;
+
         codes() noexcept
             : error_category(
                 0x809a015e2fe509bd)

--- a/include/boost/url/grammar/string_token.hpp
+++ b/include/boost/url/grammar/string_token.hpp
@@ -69,6 +69,7 @@ struct arg
     virtual char* prepare(std::size_t n) = 0;
 
     // prevent misuse
+    virtual ~arg() = default;
     arg() = default;
     arg(arg&&) = default;
     arg(arg const&) = delete;

--- a/include/boost/url/impl/error.ipp
+++ b/include/boost/url/impl/error.ipp
@@ -20,6 +20,8 @@ make_error_code(error e)
 {
     struct codes : error_category
     {
+        virtual ~codes() = default;
+
         codes() noexcept
             : error_category(
                 0xbc15399d7a4ce829)

--- a/include/boost/url/url_base.hpp
+++ b/include/boost/url/url_base.hpp
@@ -87,6 +87,7 @@ class BOOST_SYMBOL_VISIBLE
         char* old = nullptr;
     };
 
+    virtual ~url_base() noexcept = default;
     url_base() noexcept = default;
     url_base(detail::url_impl const&) noexcept;
     explicit url_base(string_view);

--- a/test/Jamfile
+++ b/test/Jamfile
@@ -17,6 +17,7 @@ project
     <toolset>msvc:<warnings-as-errors>on
     <toolset>gcc:<warnings-as-errors>on
     <toolset>clang:<warnings-as-errors>on
+    <toolset>msvc:<cxxflags>"/we4265" # Extra errors that are disabled by default
     <toolset>gcc:<cxxflags>"-Wno-unused-but-set-variable"      # Warnings in dependencies
     <toolset>clang-4:<cxxflags>"-Wno-unused-but-set-variable"  # Warnings in dependencies
     <toolset>clang-5:<cxxflags>"-Wno-unused-but-set-variable"  # Warnings in dependencies

